File: src-tauri/src/db/meta.rs
use super::*;

/// Read one value from the `meta` table (format/version/room name).
pub fn get_meta(conn: &Connection, key: &str) -> Option<String> {
    conn.query_row("SELECT value FROM meta WHERE key = ?1", [key], |r| r.get(0))
        .ok()
}

/// A2: UPSERT one value into the `meta` table — COMMANDS uses this to stamp
/// things like the embedding model/dim after a backfill.
/// CONTRACT-NOTE: `get_meta` above already existed with the exact A2 signature,
/// so only `set_meta` is new here.
pub fn set_meta(conn: &Connection, key: &str, value: &str) -> Result<(), String> {
    execute_one(
        conn,
        "INSERT INTO meta(key, value) VALUES (?1, ?2)
         ON CONFLICT(key) DO UPDATE SET value = excluded.value",
        params![key, value],
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn set_meta_upserts_and_reads_back() {
        // A2: set_meta writes, get_meta reads, a second write replaces in place.
        let conn = mem();
        assert_eq!(get_meta(&conn, "embed_model"), None);
        set_meta(&conn, "embed_model", "nomic-embed-text").unwrap();
        assert_eq!(get_meta(&conn, "embed_model").as_deref(), Some("nomic-embed-text"));
        set_meta(&conn, "embed_model", "other").unwrap();
        assert_eq!(get_meta(&conn, "embed_model").as_deref(), Some("other"));
        // UPSERT never duplicates the key.
        let n: i64 = conn
            .query_row("SELECT count(*) FROM meta WHERE key = 'embed_model'", [], |r| r.get(0))
            .unwrap();
        assert_eq!(n, 1);
    }
}

File: apps/desktop/src/main/db-host/meta.ts
/**
 * The `meta` key/value table — format/format_version/name plus anything else
 * COMMANDS stamps on later (e.g. the embedding model/dim after a backfill).
 * Ported from `src-tauri/src/db/meta.rs` (~20 lines, `get_meta`/`set_meta`).
 */

import Database from "better-sqlite3-multiple-ciphers";

/**
 * Read one value from the `meta` table, or `null` if the key is not set —
 * mirrors Rust's `query_row(...).ok()`, which collapses "no such row" (and
 * any other query failure) to `None`.
 */
export function getMeta(db: Database.Database, key: string): string | null {
  const row = db.prepare("SELECT value FROM meta WHERE key = ?").get(key) as
    | { value: string }
    | undefined;
  return row?.value ?? null;
}

/**
 * A2: UPSERT one value into the `meta` table — `get_meta` above already
 * existed with the exact A2 signature, so only `set_meta` was new on the
 * Rust side. A second call on the same key replaces the value in place
 * rather than erroring on the UNIQUE conflict or duplicating the row.
 */
export function setMeta(db: Database.Database, key: string, value: string): void {
  db.prepare(
    `INSERT INTO meta(key, value) VALUES (?, ?)
     ON CONFLICT(key) DO UPDATE SET value = excluded.value`
  ).run(key, value);
}
